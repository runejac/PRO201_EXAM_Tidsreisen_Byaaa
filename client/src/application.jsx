import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
