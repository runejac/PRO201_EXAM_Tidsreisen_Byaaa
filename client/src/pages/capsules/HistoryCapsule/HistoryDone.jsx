import React, { useContext } from "react";
import olafInfront from "../../../assets/images/olaf-infront.png";
import { CapsuleButtonGreen } from "../../../components/CapsuleButton.jsx";
import { useNavigate } from "react-router-dom";
import { User } from "../../../application.jsx";

export function HistoryDone({ id, updateToDatabase }) {
  const navigate = useNavigate();
  const { user } = useContext(User);

  const navigateToMap = () => navigate("/map");
  const navigateToMyFindings = () => {
    updateToDatabase();
    navigate("/myfindings");
  };

  // checks capsule id from database
  const capsuleNameFromDatabase = user.finishedCapsules.map((capsuleName) => {
    return capsuleName.id;
  });

  // finds capsule id matches to check whether if user has done it or not
  const filteredCapsuleNamesFromUserDatabase = capsuleNameFromDatabase.find(
    (capsuleName) => {
      return capsuleName === id;
    }
  );

  return (
    <div id={"done-container"}>
      <div className={"d-flex justify-content-center mt-3"}>
        <h1 id={"done-title"}>Fullført historiekapselen!</h1>
      </div>
      <div
        id={"olaf-infront-div"}
        className={
          "w-100 h-50 d-flex align-content-center flex-row justify-content-center gap-3"
        }
      >
        <div style={{ marginTop: "3rem" }}>
          <img
            id={"olaf-infront-img"}
            height={200}
            src={olafInfront}
            alt="bilde av olaf på fullført side"
          />
        </div>

        <div
          id={"done-text"}
          style={{ marginTop: "2rem" }}
          className={"flex-shrink-1"}
        >
          {id === filteredCapsuleNamesFromUserDatabase ? (
            <p id={"finish-paragraph"}>
              Woops! Det ser ut som du allerede har vært på denne kapselen. Gå
              til neste kapsel.
            </p>
          ) : (
            <p id={"finish-paragraph"}>
              Godt jobbet!
              <br /> Du har gjennomført en kapsel og fått poeng for det, og
              kanskje til og med lært noe nytt om Byåa!
            </p>
          )}
        </div>
      </div>
      <div
        id={"points-and-buttons-container"}
        className={
          "d-flex justify-content-center flex-column align-items-center"
        }
      >
        <div className={"mb-5"}>
          {id === filteredCapsuleNamesFromUserDatabase ? null : (
            <p id={"history-points"}>Du har fått + 20 poeng</p>
          )}
        </div>
        <div className={"d-flex flex-column"}>
          <div>
            <CapsuleButtonGreen
              onClick={navigateToMap}
              buttonText={"Tilbake til kart"}
            />
          </div>
          <div>
            <CapsuleButtonGreen
              onClick={navigateToMyFindings}
              buttonText={"Mine funn"}
            />
          </div>
        </div>
      </div>
    </div>
  );
}
