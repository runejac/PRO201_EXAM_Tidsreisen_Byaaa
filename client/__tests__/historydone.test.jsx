/**
 * @jest-environment jsdom
 */

import React from "react";
import { HistoryDone } from "../src/pages/capsules/HistoryCapsule/HistoryDone.jsx";
import { fireEvent, render } from "@testing-library/react";
import { User } from "../src/application.jsx";

//TODO: Se linje 13-14, 18, og 23 i HistoryDone.jsx

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  useNavigate: () => mockedUsedNavigate,
}));

describe("HistoryDone", () => {
  it("shows snapshot", async () => {
    const user = {
      name: "Hei",
      points: 23,
      finishedCapsules: [],
      intro: true,
      walk: false,
    };
    const component = render(
      <User.Provider value={{ user }}>
        <HistoryDone />
      </User.Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
