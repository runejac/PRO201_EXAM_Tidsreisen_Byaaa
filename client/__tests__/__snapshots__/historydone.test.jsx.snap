// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryDone shows snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="done-container"
      >
        <div
          class="d-flex justify-content-center mt-3"
        >
          <h1
            id="done-title"
          >
            Fullført historiekapselen!
          </h1>
        </div>
        <div
          class="w-100 h-50 d-flex align-content-center flex-row justify-content-center gap-3"
          id="olaf-infront-div"
        >
          <div
            style="margin-top: 3rem;"
          >
            <img
              alt="bilde av olaf på fullført side"
              height="200"
              id="olaf-infront-img"
              src="test-file-stub"
            />
          </div>
          <div
            class="flex-shrink-1"
            id="done-text"
            style="margin-top: 2rem;"
          >
            <p
              id="finish-paragraph"
            >
              Woops! Det ser ut som du allerede har vært på denne kapselen. Gå til neste kapsel.
            </p>
          </div>
        </div>
        <div
          class="d-flex justify-content-center flex-column align-items-center"
          id="points-and-buttons-container"
        >
          <div
            class="mb-5"
          />
          <div
            class="d-flex flex-column"
          >
            <div>
              <button
                class="my-2"
                id="button-capsules-green"
              >
                Tilbake til kart
              </button>
            </div>
            <div>
              <button
                class="my-2"
                id="button-capsules-green"
              >
                Mine funn
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="done-container"
    >
      <div
        class="d-flex justify-content-center mt-3"
      >
        <h1
          id="done-title"
        >
          Fullført historiekapselen!
        </h1>
      </div>
      <div
        class="w-100 h-50 d-flex align-content-center flex-row justify-content-center gap-3"
        id="olaf-infront-div"
      >
        <div
          style="margin-top: 3rem;"
        >
          <img
            alt="bilde av olaf på fullført side"
            height="200"
            id="olaf-infront-img"
            src="test-file-stub"
          />
        </div>
        <div
          class="flex-shrink-1"
          id="done-text"
          style="margin-top: 2rem;"
        >
          <p
            id="finish-paragraph"
          >
            Woops! Det ser ut som du allerede har vært på denne kapselen. Gå til neste kapsel.
          </p>
        </div>
      </div>
      <div
        class="d-flex justify-content-center flex-column align-items-center"
        id="points-and-buttons-container"
      >
        <div
          class="mb-5"
        />
        <div
          class="d-flex flex-column"
        >
          <div>
            <button
              class="my-2"
              id="button-capsules-green"
            >
              Tilbake til kart
            </button>
          </div>
          <div>
            <button
              class="my-2"
              id="button-capsules-green"
            >
              Mine funn
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
